py/dml/ctree.py

```diff
@@ -3953,13 +3953,30 @@ def mkCast(site, expr, new_type):
         old_type = safe_realtype(expr.ctype())
         if real.cmp(old_type) == 0:
             return expr
-        else:
+        elif old_type.is_arith or isinstance(old_type, TPtr):
             return mkApply(
                 expr.site,
                 mkLit(expr.site, *real.get_store_fun()),
                 (mkCast(expr.site, expr, TInt(64, False)),))
+        else:
+            raise ECAST(site, expr, new_type)
+    if (((real.is_arith or isinstance(real, TBool))
+         and (old_type.is_arith or isinstance(old_type, TBool)))
+        or ((isinstance(real, TPtr) or real.is_int or isinstance(real, TBool))
+            and (isinstance(old_type, (TPtr, TArray, TFunction))
+                 or old_type.is_int or isinstance(old_type, TBool)))):
+        assert (not (real.is_int and real.is_endian)
+                and not (old_type.is_int and old_type.is_endian))
+        return Cast(site, expr, new_type)
```

Comment:
A fairly exhaustive test in ctree_test could be appropriate?

Rather complex condition, perhaps worthwhile to split in two and duplicate `return Cast`.

Unsure if TFunction is right, if you have `local void (*p)(void);`, then isn't `cast(*p, int)` illegal?

Reply:
Also, cast between pointers and non-`uintptr_t` integers is questionable, I think they give GCC warnings. We need a stance on this.
A fairly common use case is to store an integer in a `void *`, as is seen in `uint64_event`. It can make sense to convert uint8 into such pointer, and later convert the pointer back into uint8. But I find it hard to constuct a use case where conversion from TArray into any other integer than uintptr_t would make sense (but OTOH, compatibility rules should be transitive so perhaps we must keep it a warning, like GCC does).
One option is to create a separate case for pointer-to-int conversions, produce our own warnings, and generate a double-cast via uintptr_t to silence GCC warnings.

Reply:
> A fairly exhaustive test in ctree_test could be appropriate?
> 
> Rather complex condition, perhaps worthwhile to split in two and duplicate `return Cast`.
> 
> Unsure if TFunction is right, if you have `local void (*p)(void);`, then isn't `cast(*p, int)` illegal?

My intent with allowing TFunction as old_type is so that casting a function reference to a function pointer type won't error out. As for `cast(*p, int)`, surprisingly enough, `(int) (*p)` is legal in C; the dereferenced function pointer gets automatically converted back into a pointer again. (Generates a warning though since `int` is smaller than pointer width).